File: src/main/kotlin/org/wildfly/halos/config/Environment.kt
package org.wildfly.halos.config

import mu.KotlinLoggingLevel

object Environment {
    val version: Version = Version.parse(failSafe(process.env.HALOS_VERSION) { "0.0.0" })
    val cors: Boolean = failSafe(process.env.HALOS_CORS) { true }
    val proxyUrl: String = failSafe(process.env.HALOS_PROXY_URL) { "" }
    val restVersion: String = failSafe(process.env.HALOS_REST_VERSION) { "v1" }
    val logLevel: KotlinLoggingLevel = KotlinLoggingLevel.valueOf(failSafe(process.env.HALOS_LOG_LEVEL) { "INFO" })

    private fun <T> failSafe(value: dynamic, default: () -> T): T = if (value == undefined) {
        default()
    } else {
        value.unsafeCast<T>()
    }
}

File: webpack.config.d/environment.js
const webpack = require("webpack");

if (config.mode === "development") {
    config.plugins.push(new webpack.EnvironmentPlugin({
        "HALOS_CORS": true,
        "HALOS_PROXY_URL": "http://localhost:8080",
        "HALOS_REST_VERSION": "v1",
        "HALOS_VERSION": "0.0.1"
    }))
} else {
    config.plugins.push(new webpack.EnvironmentPlugin({
        "HALOS_CORS": false,
        "HALOS_PROXY_URL": "",
        "HALOS_REST_VERSION": "v1",
        "HALOS_VERSION": "0.0.1"
    }))
}
